lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/mountpoint/codecs/YangInstanceIdentifierToPathCodec.java

```diff
@@ -42,40 +43,61 @@ public Gnmi.Path apply(final YangInstanceIdentifier path) {
         final PeekingIterator<YangInstanceIdentifier.PathArgument> iterator =
                 Iterators.peekingIterator(path.getPathArguments().iterator());
         Gnmi.PathElem.Builder previousElemBuilder = Gnmi.PathElem.newBuilder();
-
+        String previousNamespace = "";
         while (iterator.hasNext()) {
             final Gnmi.PathElem.Builder currentElemBuilder = Gnmi.PathElem.newBuilder();
             final YangInstanceIdentifier.PathArgument arg = iterator.next();
             if (doesDefineKeys(arg)) {
-                addKeysDefinitionNode((YangInstanceIdentifier.NodeIdentifierWithPredicates) arg,
-                        previousElemBuilder);
-                pathBuilder.addElem(previousElemBuilder.build());
+                if (arg instanceof YangInstanceIdentifier.NodeWithValue) {
+                    addKeysDefinitionNode((YangInstanceIdentifier.NodeWithValue) arg,
+                            previousElemBuilder);
+                } else {
+                    addKeysDefinitionNode((YangInstanceIdentifier.NodeIdentifierWithPredicates) arg,
+                            previousElemBuilder);
+                }
+                previousNamespace = getAndSetNamespaceToPath(previousElemBuilder, previousNamespace, arg);
+                pathBuilder.addElem(previousElemBuilder);
             } else {
                 addNonKeyDefinitionNode(arg, currentElemBuilder);
                 if (!doesDefineAugment(arg) && (!iterator.hasNext() || !doesDefineKeys(iterator.peek()))) {
                     /* Add this PathElem only when current path arg is not augment identifier and
                        next arg does not define key and value pairs */
-                    pathBuilder.addElem(currentElemBuilder.build());
+                    previousNamespace = getAndSetNamespaceToPath(currentElemBuilder, previousNamespace, arg);
+                    pathBuilder.addElem(currentElemBuilder);
                 }
             }
             previousElemBuilder = currentElemBuilder;
         }
-
-        // Add prefix to first path element, if requested
-        if (pathBuilder.getElemCount() > 0 && prefixFirstElement) {
-            final Gnmi.PathElem firstElement = pathBuilder.getElem(0);
-            final QName nodeType = path.getPathArguments().get(0).getNodeType();
-
-            final Optional<Module> firstElemModule = schemaContextProvider
-                    .getSchemaContext().findModule(nodeType.getNamespace(), nodeType.getRevision());
-            firstElemModule.ifPresent(module -> pathBuilder.setElem(0, firstElement.toBuilder()
-                    .setName(String.format("%s:%s", module.getName(), firstElement.getName()))));
-        }
         final Gnmi.Path resultingPath = pathBuilder.build();
         LOG.debug("Resulting gNMI Path of identifier {} is {}", path, resultingPath);
         return resultingPath;
     }
 
+    private String getAndSetNamespaceToPath(final Gnmi.PathElem.Builder elemBuilder,
```

Comment:
Please divide this method to `getModuleName` and `setPrefixToPath`. Method  `getModuleName` can be called only once after `while (iterator.hasNext())`.

Reply:
No response from author.